### PR TITLE
Initial implementation of --assembly

### PR DESCRIPTION
**Pull request by jupierce**
First part of https://issues.redhat.com/browse/ART-2940 .
Basic implementation of doozer `--assembly` parameter. It does not impact builds until group.yml is updated with
```
assemblies:
  enabled: true
```

### CODE REVIEW

File: doozerlib/runtime.py [hunk 1]
```diff
@@ -399,6 +400,14 @@ def initialize(self, mode='images', clone_distgits=True,
         self.group_dir = self.gitdata.data_dir
         self.group_config = self.get_group_config()
 
+        if self.group_config.assembles.enabled or self.enable_assemblies:
+            if re.match(r'^[a-zA-Z0-9._]+$', self.assembly) is None:
+                raise ValueError('Assembly names may only consistent of alphanumerics, ., and _. ')
```
Comment (vfreex) [inline doozerlib/runtime.py:405]: Would also like to disallow `self.assembly` to start or end with `.`:
```suggestion
            if re.fullmatch(r'[\w.]+', self.assembly) is None or self.assembly[0] == '.' or self.assembly[-1] == '.':
                raise ValueError('Assembly names may only consistent of alphanumerics, ., and _, but not start or end with a dot (.).')
```

File: doozerlib/distgit.py [hunk 2]
```diff
@@ -1514,8 +1517,14 @@ def update_distgit_dir(self, version, release, prev_release=None):
                     release = release[:-3]  # strip .p?
                     release += pval
 
+                if self.runtime.assembly:
+                    release += f'.assembly.{self.runtime.assembly}'
```
Comment (vfreex) [inline doozerlib/distgit.py:1521]: I realized this breaks the assumption that `.p[0|1]` is always at the end of the `release` field.
We need to find & replace all `endswith(".p1")` `endswith(".p0")` calls in Doozer code with `find()` or `in` statement.

**Review by vfreex — CHANGES_REQUESTED** [review]
With assemblies enabled, release field will not end with `.p0` or `.p1`. We need to replace all code that explicitly checks `.p0|1` value using the `string.endswith` call with substring find. Otherwise LGTM.

File: doozerlib/runtime.py (hunk 1 above)
Comment (jupierce) [reply doozerlib/runtime.py]: Good call. Updated.

File: doozerlib/distgit.py (hunk 2 above)
Comment (jupierce) [reply doozerlib/distgit.py:1522]: I think I caught them all - everything should be using `util.isolate_pflag_in_release` now.

**Review by sosiouxme — APPROVED** [review]
just make flake8 happy

**Review by vfreex — APPROVED** [review]
This is great! Thanks for the update.

/lgtm

**Comment by vfreex** [comment]
Merging this (Luke has approved us taking some risk and merging)

**Pull request merged**